prow/cache/cache.go

```diff
@@ -161,6 +200,11 @@ func (lruCache *LRUCache) GetOrAdd(
 		// flood of multiple requests for the same cache entry is also called
 		// "cache stampede".
 
+		// Record the cache miss.
+		if lruCache.callbacks.MissesCallback != nil {
+			lruCache.callbacks.MissesCallback(key)
+		}
```

Comment:
Consider doing this after we release the `lruCache` lock in case the callback is slow. We should try to do that for all the callbacks. 

Maybe also add a comment to the `Callbacks` struct indicating that callbacks should try to return quickly.

Reply:
Or, should we just say, "please make sure your callbacks are thread safe" and just do `go lruCache.callbacks.MissesCallback(key)` here instead"?

Reply:
I was going to suggest that initially, but for our use case (the only use case so far) that is overkill and might even hurt performance a bit.
The callbacks can start a new goroutine themselves (or pass data to a worker thread) if they need to do something heavy weight.

Reply:
Done (this was the only callback that did this AFAICS). I did this change on a new commit with the following message:

    InRepoConfigCache: do not call MissesCallback while holding the cache lock

    All other callbacks are already invoked in this style, so do it for this
    particular callback as well.

    The ForcedEvictionsCallback is invoked outside of our control, so
    technically there is nothing we can do make sure that it is also not
    being invoked while holding a lock. Apparently the lrucache
    implementation we currently
    use (https://github.com/hashicorp/golang-lru/blob/v0.5.4/simplelru/lru.go#L11)
    does not appear to use any locking (and so it not thread-safe), so we
    appear to be aligned with the "call-callback-without-holding-a-lock"
    stance.